### DISCUSSION TITLE
Jinja templates - what object properties are available?

### DISCUSSION THREAD

**Discussion by etfz** [netbox-community/netbox#12568] [Help Wanted!] [answered]
Hi,

Just got started with Jinja templating with the config templates. I'm a little confused as to what I have to work with in terms of various objects. For example, I discovered that `interface` objects has `ip_addresses`, but I could find no reference to that in the model documentation, nor is the field normally included when looking an interface object up using the API.

Is there any reference for this?

**Comment by kkthxbye-code** [comment, marked as answer]
You can check the code:

https://github.com/netbox-community/netbox/blob/b7f028fba39026ca92f439c2ba64b0dfa587d429/netbox/dcim/models/device_components.py#L670-L675

Alternatively you can print all the fields on the model you are working with in `nbshell`:

`print([f"{f.name} - {type(f).__name__}" for f in Interface._meta.get_fields()])`

```
['child_interfaces - ManyToOneRel',
 'bridge_interfaces - ManyToOneRel',
 'member_interfaces - ManyToOneRel',
 'id - BigAutoField',
 'created - DateTimeField',
 'last_updated - DateTimeField',
 'custom_field_data - JSONField',
 'device - ForeignKey',
 'name - CharField',
 'label - CharField',
 'description - CharField',
 'module - ForeignKey',
 'cable - ForeignKey',
 'cable_end - CharField',
 'mark_connected - BooleanField',
 '_path - ForeignKey',
 'enabled - BooleanField',
 'mac_address - MACAddressField',
 'mtu - PositiveIntegerField',
 'mode - CharField',
 'parent - ForeignKey',
 'bridge - ForeignKey',
 '_name - NaturalOrderingField',
 'lag - ForeignKey',
 'type - CharField',
 'mgmt_only - BooleanField',
 'speed - PositiveIntegerField',
 'duplex - CharField',
 'wwn - WWNField',
 'rf_role - CharField',
 'rf_channel - CharField',
 'rf_channel_frequency - DecimalField',
 'rf_channel_width - DecimalField',
 'tx_power - PositiveSmallIntegerField',
 'poe_mode - CharField',
 'poe_type - CharField',
 'wireless_link - ForeignKey',
 'untagged_vlan - ForeignKey',
 'vrf - ForeignKey',
 'tags - TaggableManager',
 'vdcs - ManyToManyField',
 'wireless_lans - ManyToManyField',
 'tagged_vlans - ManyToManyField',
 'ip_addresses - GenericRelation',
 'fhrp_group_assignments - GenericRelation',
 'l2vpn_terminations - GenericRelation',
 'inventory_items - GenericRelation',
 'journal_entries - GenericRelation',
 'cable_terminations - GenericRelation',
 'tagged_items - GenericRelation']
```

**Reply by etfz** [reply]
I'll check it out. Thank you!

**Reply by candlerb** [reply]
Also in nbshell, you can try `help(Interface)`.  You'll then get some help text for each one, although in many cases it's generic Djangoese.

```
 |  ip_addresses
 |      Accessor to the related objects manager on the one-to-many relation created
 |      by GenericRelation.
 |
 |      In the example::
 |
 |          class Post(Model):
 |              comments = GenericRelation(Comment)
 |
 |      ``post.comments`` is a ReverseGenericManyToOneDescriptor instance.
```

**Comment by robertmain** [comment]
Seconding this - I'm having a hard time writing my config templates. How does one get the IP address for an interface (for example)?

Is it: 

```jinja2
{{ interface.address }}
```

or 

```jinja2
{{ interface.ip_address.address }}
```

or

```jinja2
{{ interface.ip_address.value }}
```

or


```jinja2
{{ interface.ip_address.ip }}
```

or is it

```jinja2
{{ interface.ip_address | first }}
```

I ask in earnest because I genuinely can't tell (even after using nbshell, I'm a little unused to Django, sorry). If someone has any insight, I'd appreciate it.

**Comment by zndrr** [comment]
I too have just started out on serious template construction and seeking guidance.
Will check through the above for now, but definitely keeping an eye on this space.

**Comment by candlerb** [comment]
> How does one get the IP address for an interface (for example)?

In general, an interface has multiple IP addresses, so you could either get the first one, or all of them.

Here's a custom link I use for getting the management IP address of a device (which will be an explicit attribute in NB 3.6):

```
# Link text
{% if object.interfaces.filter(mgmt_only=True).count() == 1 and object.interfaces.get(mgmt_only=True).ip_addresses.count() == 1 %}Open IPMI{% endif %}

# Link URL
https://{{ object.interfaces.get(mgmt_only=True).ip_addresses.first().address.ip }}/
```

In nbshell:

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
...
>>> d = Device.objects.get(name="nuc1")
>>> i = d.interfaces.first()
>>> [a for a in dir(i) if "address" in a]
['count_ipaddresses', 'ip_addresses', 'mac_address']
>>> [a for a in i.ip_addresses.all()]
[<IPAddress: 10.12.255.11/24>, <IPAddress: 2001:db8::11/64>]
>>> [a.address for a in i.ip_addresses.all()]
[IPNetwork('10.12.255.11/24'), IPNetwork('2001:db8::11/64')]
>>> [a.address.ip for a in i.ip_addresses.all()]
[IPAddress('10.12.255.11'), IPAddress('2001:db8::11')]
>>> [str(a.address) for a in i.ip_addresses.all()]
['10.12.255.11/24', '2001:db8::11/64']
>>> [str(a.address.ip) for a in i.ip_addresses.all()]
['10.12.255.11', '2001:db8::11']
```

Here, `a` is a Netbox IPAddress object, `a.address` is the IPNetwork object with the prefix length, and `a.address.ip` is just the  address without prefix.

**Reply by robertmain** [reply]
Interesting, thanks. How did you work this out? I tried looking at [the data model](https://docs.netbox.dev/en/stable/models/ipam/ipaddress/) (before I saw your reply) and put this together:

```jinja2
{% set physical = device.interfaces.all() | selectattr('type', 'ne', 'virtual') %}
{% for interface in physical %}
  set interfaces ethernet {{ interface.name }} address {{ interface.address }} 
  set interfaces ethernet {{ interface.name }} description {{ interface.label }}
  set interfaces ethernet {{ interface.name }} duplex auto
  set interfaces ethernet {{ interface.name }} speed {{ interface.speed }}
{% endfor %}
```

That gave me this error:

```
Traceback (most recent call last):
  File "/opt/netbox/netbox/dcim/views.py", line 2087, in get_extra_context
    rendered_config = config_template.render(context=context_data)
  File "/opt/netbox/netbox/extras/models/configs.py", line 252, in render
    output = template.render(**context)
  File "/opt/netbox/venv/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/netbox/venv/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 3, in top-level template code
  File "/opt/netbox/venv/lib/python3.10/site-packages/jinja2/sandbox.py", line 326, in getattr
    value = getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'dcim.models.device_components.Interface object' has no attribute 'address'
```

which (given what you said about an interface having multiple addresses kinda makes sense I guess....

What I'm essentially getting at here is what you looked at or read to figure out the above snippet. I'm trying to put together a config template for my ER4 where a physical interface may have multiple IPs, but at the same time each physical interface may have VLANs hanging off it which then have exactly one IP each(generally just as the default gateway IP, not for management per-se).

**Reply by robertmain** [reply]
Right now I'm trying things out and having to save and re-save the config template which works, but it's a tad slow for experimenting with things....I'd be curious to know if there's a REPL environment out there somewhere to test stuff out in kinda like the one in HomeAssistant. 

![image](https://github.com/netbox-community/netbox/assets/537684/74d536d6-decb-4708-a964-b904aec808c8)

Not necessarily a full GUI application like that, even a shell would be cool. I've tried playing with around with nbshell, but that feels like it's very...open-ended and unrestricted(if that makes sense), so an expression that works in nbshell, may not necessarily work in the jinja2 config template. Hopefully that makes sense.

**Reply by kkthxbye-code** [reply]
> which (given what you said about an interface having multiple addresses kinda makes sense I guess....

It's right there in the answer by candlerb. You did:

`interface.address`

He did:

`interface.ip_addresses.first().address.ip`

If you need it with the mask, remove the `.ip` from the end. Note that the above is likely to produce wrong results if the interface has no IP. You would need to filter the interfaces or do a conditional.

**Reply by robertmain** [reply]
> It's right there in the answer by candlerb. You did:

@kkthxbye-code yeah, that's what I meant when I added _"before I saw your reply"_ in the parens :) - what I'm effectively getting at here is: "Is there some master document somewhere that outlines what objects have what properties in the jinja templates?" It's possible that the answer is just "Go look at the Django data model and use those" which I did and it didn't work, but that may be because I just didn't quite understand it properly.

> If you need it with the mask, remove the .ip from the end. Note that the above is likely to produce wrong results if the interface has no IP. You would need to filter the interfaces or do a conditional.

What kind of wrong results do you mean? Like `Undefined` or something? Most of my interfaces (possibly all of them) have IPs, but I'll probably check for this as you suggested for completeness and robustness.

**Reply by kkthxbye-code** [reply]
It's not a great answer, but it's the best we got. I made an earlier reply in this very topic explaining the example with IP addresses:

https://github.com/netbox-community/netbox/discussions/12568#discussioncomment-5881716

Experimenting in nbshell is your best bet. Everything you can do there should be possible in jinja as long as it's possible in jinja syntax. If you find yourself doing complicated stuff often, it might be best to write a custom filter:

https://docs.netbox.dev/en/stable/configuration/system/#jinja2_filters

A playground for quickly testing configuration templates is a good idea, but not something we currently support. It should be easily doable in a plugin. Not sure if it's something we want built-in though, some of the other maintainers would have to weigh in. Making a feature request for it would be fine though.

> What kind of wrong results do you mean? Like Undefined or something? Most of my interfaces (possibly all of them) have IPs, but I'll probably check for this as you suggested for completeness and robustness.

I think it just renders nothing if there's no IP, but maybe it actually throws an exception, not sure. You'll have to play with it. If you always have IPs on all physical interfaces it should be fine.

**Reply by robertmain** [reply]
Thanks for the help @kkthxbye-code - I ended up with this:

```jinja2
{%- set interfaces = device.interfaces.all() -%}
{%- set physical = interfaces | selectattr('type', 'ne', 'virtual') -%}
{%- for interface in physical -%}
{%- if interface.ip_addresses.first() %}
set interfaces ethernet {{ interface.name }} address {{ interface.ip_addresses.first().address.ip }}
{%- endif %}
{%- if interface.label %}
set interfaces ethernet {{ interface.name }} description {{ interface.label }}
{%- endif %}
set interfaces ethernet {{ interface.name }} duplex auto
set interfaces ethernet {{ interface.name }} speed {{ interface.speed | default('auto') }}
{% endfor %}
```

**Reply by zndrr** [reply]
This has actually been hugely educational! Made strides in templating via this discussion.
Would there be any possibility of adding some breadcrumbs like the above to general docos? Some of us just need to be guided in the right direction.

I'll mirror **robertmain** on the thanks. Much appreciated.

**Reply by kkthxbye-code** [reply]
> Would there be any possibility of adding some breadcrumbs like the above to general docos?

I think we are always open to modify the docs, but it's hard with stuff like custom scripts, configuration templates and nbshell as you can do pretty much anything. They are inherently power user features, and require some sort of knowledge about the internal structure of the models in netbox and the django ORM.

In my opinion stuff like that is more suited for blog posts/how-to guides, but there might be room for some very limited pointers.

**Reply by zndrr** [reply]
@kkthxbye-code I think the pointers here would be sufficient to share? Particularly the discovery stuff.
That was enough for me at least, and it seems robertmain also; could intuit a lot from that.

Of course I'm not out of the woods yet, so a lot is left to trial-and-error. Juggling J2 whitespacing and readability is the same regardless XD.
For now, what I have mocked up should be sufficient as PoC for my job, since config rendering won't be relied upon in the short term.

**Comment by candlerb** [comment]
> Interesting, thanks. How did you work this out?

Firstly, in nbshell there is info on the attributes on an object using `help(MODEL)` - it includes comments extracted from the source code and/or from Django generic documentation.

e.g. `help(Interface)` includes the following within the results:

```
 |  ip_addresses
 |      Accessor to the related objects manager on the one-to-many relation created
 |      by GenericRelation.
 |
 |      In the example::
 |
 |          class Post(Model):
 |              comments = GenericRelation(Comment)
 |
 |      ``post.comments`` is a ReverseGenericManyToOneDescriptor instance.
```

Secondly, in Python in general, you can get a list of all the property names and method names on an object using `dir(obj)`.  So for example:

```
>>> i = d.interfaces.first()
>>> dir(i)
```

will list *everything* on that particular interface object - which will be a huge list (it includes all the methods inherited from parent classes, and lots of private methods methods).

If you know roughly what you're looking for, you can filter this list.  I showed:

```
>>> [a for a in dir(i) if "address" in a]
```

That's a list comprehension: it take objects from a list (`for a in dir(i)`), filters it (`if "address" in a` which in this case is a substring match) and then returns a new list comprising those elements (the first `a` after the opening bracket is the expression used to create each new element in the list).

In both ways you learn that `ip_addresses` is an available property or method on Interface, but `address` is not. Hence when you wrote

```
  set interfaces ethernet {{ interface.name }} address {{ interface.address }} 
```

you got a message telling you exactly that:

```
jinja2.exceptions.UndefinedError: 'dcim.models.device_components.Interface object' has no attribute 'address'
```

And finally, there's the source code for [Interface](https://github.com/netbox-community/netbox/blob/master/netbox/dcim/models/device_components.py#L540-L692) and [IPAddress](https://github.com/netbox-community/netbox/blob/master/netbox/ipam/models/ip.py#L670-L748). On Interface, this gives you:

```
    ip_addresses = GenericRelation(
        to='ipam.IPAddress',
        content_type_field='assigned_object_type',
        object_id_field='assigned_object_id',
        related_query_name='interface'
    )
```

But that involves getting more into Django internals.

**Reply by zndrr** [reply]
Sorry @candlerb I can't for the life of me figure out how to distinguish ip4 and ip6 address families. I can distinguish primary_ip(4/6) due to direct device relationship, but have so far failed to discover syntax for non-mgmt addressing.
Context is rendering ip4 and ip6 addressing distinctly on an interface template. Any chance you could help there? TIA

**Reply by robertmain** [reply]
@candlerb very informative, thanks!

**Reply by candlerb** [reply]
> I can't for the life of me figure out how to distinguish ip4 and ip6 address families

```
>>> help(IPAddress)
...
 |  ----------------------------------------------------------------------
 |  Readonly properties defined here:
 |
 |  family
 |
 |  ----------------------------------------------------------------------
...

>>> a = Device.objects.get(name="nuc1").interfaces.first().ip_addresses.first()
>>> a.address
IPNetwork('10.12.255.11/24')
>>> a.family
4
```

**Reply by zndrr** [reply]
Right, seems I didn't go deep enough -_-;; , and also missed that when checking the help previously...

I can see the info in nbshell, so it's just stringing it together in J2. first() and last() give you the choice but not all()... makes sense, but I just have to learn how to filter specifically. My lack of overall understanding of the language is showing.

Skill issue basically. Sorry and thanks, I can brute force my way there. For now need a to take a break.

**Reply by candlerb** [reply]
`first()` and `last()` give you a single concrete object, or else `None` - that's easy to handle.

`all()` gives you a proxy manager object. I think you should be able to iterate over it: `{%- for interface in blah.all() %}` - but if not, you can always turn it into a concrete list of objects using `| list` as @robertmain showed.

The Jinja2 filter `foo | list` does the same as `list(foo)` in Python.

It is indeed annoying that Jinja2 is kind-of Python-ish, but different :-(

**Reply by ross-cello** [reply]
Cool, I freaking got it. Thanks heaps @candlerb (I'm the same guy as above).
Abridged mock-up below. Subject to proper cleanup, whitespacing etc. Thought it might have value including here for others.

```
{% for interface in device.interfaces.all() %}
!
interface {{ interface.name }}
{% if interface.description %} description {{ interface.description }}{% else %} no description{% endif %}

{% for ip in interface.ip_addresses.all() %}
{%- if ip.family ==  4 %} ip address {{ ip.address.ip }} {{ ip.address.netmask }}{% endif %}
{%- if ip.family ==  6 %} ipv6 address {{ ip.address }}{% endif %}
{% endfor -%}
exit
{% endfor -%}
```

**Comment by robertmain** [comment]
I don't have a ton to contribute, but I wanted to check back in with my working template so far...hopefully it helps somebody...


```jinja2


{%- set vlans = device.site.vlans.all() | list -%}
{%- set interfaces = device.interfaces.all() -%}
{%- set physical = interfaces | selectattr('type', 'ne', 'virtual') | list -%}
{%- set virtual = interfaces | selectattr('type', 'eq', 'virtual') | list -%}
{%- set port_speeds = {
} -%}
{%- for interface in physical -%}
{%- if interface.ip_addresses.first() %}
set interfaces ethernet {{ interface.name }} address {{ interface.ip_addresses.first().address }}
{%- endif %}
{%- if interface.label %}
set interfaces ethernet {{ interface.name }} description {{ interface.label }}
{%- endif %}
{%- if not interface.enabled %}
set interfaces ethernet {{ interface.name }} disable
{%- endif %}
set interfaces ethernet {{ interface.name }} duplex auto
set interfaces ethernet {{ interface.name }} speed {{ port_speeds[interface.speed | string] or 'auto' }}
{%- if interface.mtu %}
set interfaces ethernet {{ interface.name }} mtu {{ interface.mtu }}
{%- endif %}
{%- for vif in virtual | selectattr('parent.id', 'eq', interface.id) | list %}
{% set vlan = vlans | selectattr('id', 'eq', vif.untagged_vlan_id) | first %}
{%- if vif.ip_addresses.first() %}
set interfaces ethernet {{ vif.parent.name }} vif {{ vlan.vid }} address {{ vif.ip_addresses.first().address }}
{%- endif %}
set interfaces ethernet {{ vif.parent.name }} vif {{ vlan.vid }} description {{ vlan.name }}
{%- if vif.mtu %}
set interfaces ethernet {{ vif.parent.name }} vif {{ vlan.vid }} mtu {{ vif.mtu }}
{%- endif %}
{%- endfor %}
{% endfor %}
```


It looks a bit of a wall of text so I'm sure there's a neater way to express that ^ 

Critique welcome.

**Reply by zndrr** [reply]
I have something for interfaces that does IP addressing. It's stupidly long though, so I'm sure it can be refined further.
It accounts for the number of IP's, being 0, 1 or 2+, and sets appropriate commands.
I wanted to be explicit in rendering syntax where data is not present.

My background is ISP, so there are a lot of different combinations of configuration.
Created on v3.6-beta2. Haven't tested on v3.5.x.

```
{% for intf in device.interfaces.all() %}
{%- set ips = intf.ip_addresses.count() -%}
!
interface {{ intf.name }}
{% if intf.enabled %} no shutdown{% else %} shutdown{% endif %}
{% if intf.description %} description {{ intf.description }}{% else %} no description{% endif %}
{% if intf.mtu %} mtu {{ intf.mtu }}{% else %} no mtu{% endif %}
{%- if ips == 0 %}
 no ip address
 no ipv6 address{% endif %}
{% for ip in intf.ip_addresses.all() %}{%- set ip4, ip6 = ip.address.version == 4, ip.address.version == 6 -%}
{% if ips >= 2 %}
{%- if ip4 %} ip address {{ ip.address.ip }} {{ ip.address.netmask }}{% endif %}
{%- if ip6 %} ipv6 address {{ ip.address }}{% endif %}
{% elif ips == 1 %}
{%- if ip4 %} ip address {{ ip.address.ip }} {{ ip.address.netmask }}
 no ipv6 address
{% elif ip6 %} no ip address
 ipv6 address {{ ip.address }}
{% endif -%}{% endif %}{%- endfor -%}{% endfor -%}
!
```
Would render something like below. Bearing in mind, it still lacks ipv6 enable command and ip4/ip6 mtu. That would need to be done in isolation to prevent duplicity.
```
!
interface vi0/0.10
 no shutdown
 no description
 no mtu
 no ip address
 no ipv6 address
!
interface vi0/1
 shutdown
 no description
 no mtu
 ip address 1.1.1.3 255.255.255.255
 ipv6 address fe83::3/112
 ipv6 address fe83::1:3/112
!
interface lo0
 no shutdown
 description MGMT
 no mtu
 ip address 1.1.1.1 255.255.255.255
 ipv6 address fe80::1/64
!
interface lo1
 no shutdown
 no description
 mtu 1567
 ip address 1.1.1.2 255.255.255.255
 no ipv6 address
!
interface lo2
 no shutdown
 no description
 no mtu
 no ip address
 ipv6 address fe82::1/64
!
```

**Reply by candlerb** [reply]
I think that won't work as expected if you have 2 ipv6 addresses but no ipv4 address. I'd do something like this (untested):

```
!
interface {{ intf.name }}
{%- for ip in intf.ip_addresses.filter(family=4).all() %}
 ip address {{ ip.address.ip }} {{ ip.address.netmask }}{% if not loop.first %} secondary{% endif %}
{%- else %}
 no ip address
{%- endfor %}
{%- for ip in intf.ip_addresses.filter(family=6).all() %}
 ipv6 address {{ ip.address }}
{%- else %}
 no ipv6 address
{%- endfor %}
```

See [jinja2 "for" loop documentation](https://jinja.palletsprojects.com/en/3.0.x/templates/#for)

**Reply by zndrr** [reply]
You're right, it doesn't quite "work" as desired (just tested). Also have more whitespacing to mend.
Good call, I hadn't considered such a configuration and my early runs, but I know exactly why.

```
!
interface lo4
 no shutdown
 no description
 no mtu
 ip address 4.4.4.1 255.255.255.255 ip address 4.4.4.2 255.255.255.255 ip address 4.4.4.3 255.255.255.255
 load-interval 30
!
interface lo6
 no shutdown
 no description
 no mtu

 ipv6 address fe86::1/128
 ipv6 address fe86::2/128
 ipv6 address fe86::3/128
 load-interval 30

!
```

Yours does look much cleaner and concise.
I have played around a bit to refine, but wasn't yet able to discover a method like you've expressed. Lots of trial and error, and a sprinkle of nbshell.
It's good that you covered the ipv4 secondary use case too - that was one of the limitations I had noted in my method.
Might give that a go and factor it in. Thanks for your input.

**Reply by zndrr** [reply]
Ah, perhaps not:

```
<class 'django.core.exceptions.FieldError'>

Cannot resolve keyword 'family' into field. Choices are: address, assigned_object, assigned_object_id, assigned_object_type, assigned_object_type_id, bookmarks, comments, created, custom_field_data, description, dns_name, fhrpgroup, id, interface, journal_entries, last_updated, nat_inside, nat_inside_id, nat_outside, role, services, status, tagged_items, tags, tenant, tenant_id, vminterface, vrf, vrf_id
```

'family' was one item in particular that I did struggle with a bit. But I at least have something to play with.

**Reply by candlerb** [reply]
I did say untested :-) It turns out "family" is a virtual property on the IPAddress model: you can't directly filter it with a Django filter. (I think it used to be a concrete database column, but not any more)

I expect you can use a [jinja2 filter](https://jinja.palletsprojects.com/en/3.1.x/templates/#jinja-filters.selectattr) though: perhaps something like this

```
intf.ip_addresses.all() | selectattr("family","equalto",4)    ## edited and corrected
```

**Reply by ross-cello** [reply]
I've played around a bit to no success, both with 'family' and 'version'. Mostly complains about iteration.
With loop.first, syntax can be somewhat shortened. But since it runs against ip4 and ip6 (inside ip_addresses.all() ), it'll only apply against first occurrence, which typically is ip4 when both families are present... that does make sense.
So far starting loops with ip4 and ip6 isolated has been unsuccessful. What a journey so far... XD

**Reply by candlerb** [reply]
I gave you the wrong syntax for selectattr. Here it is again, this time tested and working.

```
hostname {{ device.name }}
{%-for intf in device.interfaces.all() %}
!
interface {{ intf.name }}
{%- for ip in intf.ip_addresses.all() | selectattr("family","equalto",4) %}
 ip address {{ ip.address.ip }} {{ ip.address.netmask }}{% if not loop.first %} secondary{% endif %}
{%- else %}
 no ip address
{%- endfor %}
{%- for ip in intf.ip_addresses.all() | selectattr("family","equalto",6)  %}
 ipv6 address {{ ip.address }}
{%- else %}
 no ipv6 address
{%- endfor %}
{%- endfor %}
```

**Reply by kkthxbye-code** [reply]
Here's how you filter by family in a queryset:

```
>>> IPAddress.objects.filter(address__family=4)
<RestrictedQuerySet [<IPAddress: 10.0.0.1/24>]>
>>> IPAddress.objects.filter(address__family=6)
<RestrictedQuerySet [<IPAddress: 2001:4860:4860::8888/126>]>
```

**Reply by candlerb** [reply]
Thanks. Confirmed that works too:

```
hostname {{ device.name }}
{%-for intf in device.interfaces.all() %}
!
interface {{ intf.name }}
{%- for ip in intf.ip_addresses.filter(address__family=4) %}
 ip address {{ ip.address.ip }} {{ ip.address.netmask }}{% if not loop.first %} secondary{% endif %}
{%- else %}
 no ip address
{%- endfor %}
{%- for ip in intf.ip_addresses.filter(address__family=6) %}
 ipv6 address {{ ip.address }}
{%- else %}
 no ipv6 address
{%- endfor %}
{%- endfor %}
```

**Reply by ross-cello** [reply]
Ugh dammit, I tried permutations of both `selectattr("family","equalto",6)` and `intf.ip_addresses.filter(address__family=4)` but didn't get the syntax quite right, ending in failure! Still boils down to an issue of education. Thanks again for delivering.
I'll edit above once I refined. Needed to jig a bit to factor in first loop ip mtu and ip6 enable

```
{% for intf in device.interfaces.all() %}
!
interface {{ intf.name }}
{%- for ip in intf.ip_addresses.filter(address__family=4) %}
{%- if loop.first %}
 ip mtu 1500
 !--- add other single occurrence ip4 commands here
 ip address {{ ip.address.ip }} {{ ip.address.netmask }}{% else %}
 ip address {{ ip.address.ip }} {{ ip.address.netmask }} secondary{% endif %}
{%- else %}
 no ip address
{%- endfor %}
{%- for ip in intf.ip_addresses.filter(address__family=6) %}
{%- if loop.first %}
 ipv6 enable
 ipv6 mtu 1500
 !--- add other single occurrence ip6 commands here
 ipv6 address {{ ip.address }}{% else %}
 ipv6 address {{ ip.address }}{% endif %}
{%- else %}
 no ipv6 address
{%- endfor %}
 load-interval 30
{% endfor %}
!
```
I'm okay with that level of duplicity. It actually resembles typical syntax for secondaries. Please go easy on me if you see opportunities to refine futher. XD

**Comment by mtinberg** [comment]
Yeah, the Swagger docs are awesome for listing out every query parameter which is a great reference, but I don't know of an easy way to list all the Django object filters as they are used in Jinja2 templates (or graphql).  There are some hints here

https://demo.netbox.dev/static/docs/administration/netbox-shell/
https://docs.djangoproject.com/en/4.2/ref/models/querysets/#id4

which is great, but having a lookup table for reference would be easier instead of having to use my slow brain to find this doc and calculate what the filter should look like for a particular object type.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by ross-cello** [reply]
Dude this is awesome; the second link in particular (first doesn't load for me).
Didn't think to look for Django resources before. Now I can filter for dot1q etc:

```
>>> a.interfaces.filter(name__iregex=r'\.')
<RestrictedQuerySet [<Interface: GigabitEthernet0/0/0.10>, <Interface: GigabitEthernet0/0/1.20>]>
```

Edit: Constructed this quickly (subject to refinement):

```
{% for ifd1q in device.interfaces.filter(name__contains=".") %}
{% set dot1q = ifd1q.name.rsplit('.', 1) %}
!
interface {{ ifd1q.name }}
 encapsulation dot1q {{ dot1q[1] }}
{% endfor %}
!
```
Rendering this
```
!
interface GigabitEthernet0/0/0.10
 encapsulation dot1q 10
!
interface GigabitEthernet0/0/1.20
 encapsulation dot1q 20
!
```

**Comment by bitcollector1** [comment]
This is an amazing thread, thanks to all the fine folks who contributed to this discussion. 

I was able to get my templates built completely a few weeks after the feature was released (with some help from slack forum) but I've learned a lot by just reading this thread.
